File: src/main/java/com/lessmarkup/engine/filesystem/TemplateContext.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.filesystem;

import com.lessmarkup.Constants;
import com.lessmarkup.framework.helpers.LanguageHelper;
import com.lessmarkup.framework.helpers.StringHelper;
import com.lessmarkup.framework.system.RequestContextHolder;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.system.RequestContext;
import com.lessmarkup.interfaces.system.ResourceCache;
import com.lessmarkup.interfaces.system.SiteConfiguration;
import com.samskivert.mustache.Mustache;
import java.io.StringReader;

public class TemplateContext {
    
    private final Mustache.Lambda text;
    private final Mustache.Lambda property;
    private final Mustache.Lambda map;
    
    public static Mustache.Compiler createCompiler(ResourceCache resourceCache) {
        return Mustache.compiler()
                .escapeHTML(false)
                .withDelims("[[ ]]")
                .withLoader(s -> new StringReader(resourceCache.readText(s)));
    }
    
    public TemplateContext(DataCache dataCache) {
        this.text = (frag, out) -> {
            String key = frag.execute();
            out.write(LanguageHelper.getText(Constants.ModuleType.MAIN, key));
        };
        this.property = (frag, out) -> {
            String key = StringHelper.toJsonCase(frag.execute());
            String value;
            switch (key) {
                case "rootPath":
                    value = RequestContextHolder.getContext().getRootPath();
                    break;
                default:
                    value = dataCache.get(SiteConfiguration.class).getProperty(key);
                    break;
            }
            out.write(value);
        };
        RequestContext requestContext = RequestContextHolder.getContext();
        this.map = (frag, out) -> {
            String key = frag.execute();
            String value = requestContext.mapPath(key);
            out.write(value);
        };
    }

    public Mustache.Lambda getText() {
        return text;
    }

    public Mustache.Lambda getProperty() {
        return property;
    }
    
    public Mustache.Lambda getMap() {
        return map;
    }
}

File: src/main/scala/com/lessmarkup/engine/filesystem/TemplateContext.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.filesystem

import java.io.{Reader, StringReader, Writer}

import com.lessmarkup.Constants
import com.lessmarkup.framework.helpers.{LanguageHelper, StringHelper}
import com.lessmarkup.framework.system.RequestContextHolder
import com.lessmarkup.interfaces.cache.DataCache
import com.lessmarkup.interfaces.system.{ResourceCache, SiteConfiguration}
import com.samskivert.mustache.Mustache.{Lambda, TemplateLoader}
import com.samskivert.mustache.{Mustache, Template}

object TemplateContext {
  def createCompiler(resourceCache: ResourceCache): Mustache.Compiler = {
    Mustache.compiler.escapeHTML(false).withDelims("[[ ]]").withLoader(
      new TemplateLoader {
        override def getTemplate(s: String): Reader = {
          new StringReader(resourceCache.readText(s).getOrElse(""))
        }
      })
  }
}

class TemplateContext(dataCache: DataCache) {

  private final val text: Mustache.Lambda = new Lambda {
    override def execute(fragment: Template#Fragment, writer: Writer): Unit = {
      val key = fragment.execute()
      val text = LanguageHelper.getText(Constants.ModuleTypeMain, key)
      writer.write(text)
    }
  }

  private final val property: Mustache.Lambda = new Lambda {
    override def execute(fragment: Template#Fragment, writer: Writer): Unit = {
      val key = StringHelper.toJsonCase(fragment.execute())
      val value = key match {
        case "rootPath" => RequestContextHolder.getContext.getRootPath
        case _ => dataCache.get(classOf[SiteConfiguration]).getProperty(key)
      }
      writer.write(value)
    }
  }

  private final val map: Mustache.Lambda = new Lambda {
    override def execute(fragment: Template#Fragment, writer: Writer): Unit = {
      val key = fragment.execute()
      val value = RequestContextHolder.getContext.mapPath(key)
      writer.write(value)
    }
  }

  def getText: Mustache.Lambda = text
  def getProperty: Mustache.Lambda = property
  def getMap: Mustache.Lambda = map
}